File: kotlin/packages/cli-function/src/main/kotlin/com/prog/cli/function/Greeter.kt
package com.prog.cli.function

import com.prog.goodbye.world.printGoodbye
import com.prog.hello.world.printGreeting

object Greeter {
    /**
     * # `main`
     *
     * This function is the entry point of the application.
     *
     * __Arguments:__
     *
     * * `args: Array<String>` - Command line arguments
     */
    @JvmStatic
    fun main(args: Array<String>) {
        println("===================================================")
        printWelcomeMessage(args)
        printGoodbyeMessage(args)
        println("===================================================")
    }

    /**
     * # `printWelcomeMessage`
     *
     * This function prints the welcome message by calling the `printGreeting`
     *
     * __Arguments:__
     *
     * * `args: Array<String>`
     */
    fun printWelcomeMessage(args: Array<String>) {
        println(printGreeting(args))
    }

    /**
     * # `printGoodbyeMessage`
     *
     * This function prints the goodbye message by calling the `printGoodbye`
     *
     * __Arguments:__
     *
     * * `args: Array<String>`
     */
    fun printGoodbyeMessage(args: Array<String>) {
        System.err.println(printGoodbye(args))
    }
}

File: haskell/src/CLI/Function/Greeter.hs
module CLI.Function.Greeter (printWelcomeMessage, printGoodbyeMessage) where

import System.IO (hPutStrLn, stderr)

import Goodbye.World (printGoodbye)
import Hello.World (printGreeting)

-- |
-- = @printWelcomeMessage@
--
-- The function @printWelcomeMessage@ is used to print the welcome message.
--
-- __Arguments:__
--
-- * @args: [String]@ - The command line arguments.
--
-- __Return:__
--
-- * @IO ()@ - The welcome message.
printWelcomeMessage :: [String] -> IO ()
printWelcomeMessage args = do
    putStrLn $ printGreeting args

-- |
-- = @printGoodbyeMessage@
--
-- The function @printGoodbyeMessage@ is used to print the goodbye message.
--
-- __Arguments:__
--
-- * @args: [String]@ - The command line arguments.
--
-- __Return:__
--
-- * @IO ()@ - The goodbye message.
printGoodbyeMessage :: [String] -> IO ()
printGoodbyeMessage args = do
    hPutStrLn stderr $ printGoodbye args
